File: backend/src/deaths_door/script_name.py
from __future__ import annotations

from enum import Enum


class ScriptName(str, Enum):
    """The name of a script."""

    TROUBLE_BREWING = "trouble_brewing"
    SECTS_AND_VIOLETS = "sects_and_violets"
    BAD_MOON_RISING = "bad_moon_rising"

    def __str__(self) -> str:
        """Return the human readable name of the script."""
        match self:
            case ScriptName.TROUBLE_BREWING:
                return "Trouble Brewing"
            case ScriptName.SECTS_AND_VIOLETS:
                return "Sects and Violets"
            case ScriptName.BAD_MOON_RISING:
                return "Bad Moon Rising"

    @classmethod
    def from_str(cls, name: str) -> ScriptName | None:
        """Return the ScriptName for a given string if present, else return none."""
        for script in cls:
            if script.value == name.lower():
                return script

File: backend/src/script_name.rs
//! The name of a script (edition).

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ScriptName {
    TroubleBrewing,
    SectsAndViolets,
    BadMoonRising,
}

impl ScriptName {
    pub const ALL: [ScriptName; 3] = [
        ScriptName::TroubleBrewing,
        ScriptName::SectsAndViolets,
        ScriptName::BadMoonRising,
    ];

    /// Machine-readable value (e.g. `trouble_brewing`).
    pub fn value(&self) -> &'static str {
        match self {
            ScriptName::TroubleBrewing => "trouble_brewing",
            ScriptName::SectsAndViolets => "sects_and_violets",
            ScriptName::BadMoonRising => "bad_moon_rising",
        }
    }

    /// Human-readable display name.
    pub fn display(&self) -> &'static str {
        match self {
            ScriptName::TroubleBrewing => "Trouble Brewing",
            ScriptName::SectsAndViolets => "Sects and Violets",
            ScriptName::BadMoonRising => "Bad Moon Rising",
        }
    }

    /// Parse from a value string, case-insensitively.
    #[allow(clippy::should_implement_trait)]
    pub fn from_str(name: &str) -> Option<ScriptName> {
        let lower = name.to_lowercase();
        ScriptName::ALL.into_iter().find(|s| s.value() == lower)
    }
}
